File: web/src/pipeline/mapping.ts
/**
 * Plane -> sphere parameterization (Lambert azimuthal equal-area).
 * Port of ball_stencil/mapping.py.
 *
 *   rho = |p - center|
 *   u   = (rho / R_ref) * sin(cap/2)        # in [0,1]
 *   phi = 2 * asin(u)                       # polar angle from pole
 *   theta = atan2(dy, dx)                   # azimuth (dy flipped if flip_v)
 *   P(r) = r * (sin phi cos theta, sin phi sin theta, cos phi)
 */

export class Mapper {
  readonly cx: number;
  readonly cy: number;
  readonly rRef: number;
  readonly capAngleRad: number;
  readonly flipV: boolean;

  constructor(
    center: [number, number],
    rRef: number,
    capAngleRad: number,
    flipV: boolean,
  ) {
    if (!(rRef > 0.0)) {
      throw new Error(
        `design reference radius must be > 0 (got ${rRef}); the artwork may ` +
          "have collapsed to a single point or an empty region",
      );
    }
    this.cx = center[0];
    this.cy = center[1];
    this.rRef = rRef;
    this.capAngleRad = capAngleRad;
    this.flipV = flipV;
  }

  get sinHalfCap(): number {
    return Math.sin(this.capAngleRad / 2.0);
  }

  /** Unit sphere direction for one planar point. */
  direction(x: number, y: number): [number, number, number] {
    const dx = x - this.cx;
    const dyRaw = y - this.cy;
    const dy = this.flipV ? -dyRaw : dyRaw;
    const rho = Math.hypot(dx, dy);
    const theta = Math.atan2(dy, dx);
    let u = (rho / this.rRef) * this.sinHalfCap;
    if (u < 0) u = 0;
    else if (u > 1) u = 1;
    const phi = 2.0 * Math.asin(u);
    const sphi = Math.sin(phi);
    return [sphi * Math.cos(theta), sphi * Math.sin(theta), Math.cos(phi)];
  }

  /** d(arc length)/d(rho) at planar radius rho. */
  radialScale(rho: number, radius: number): number {
    const s = this.sinHalfCap / this.rRef;
    let u = rho * s;
    if (u < 0.0) u = 0.0;
    else if (u > 1.0 - 1e-12) u = 1.0 - 1e-12;
    const dphiDrho = (2.0 * s) / Math.sqrt(1.0 - u * u);
    return radius * dphiDrho;
  }

  /** Arc length per SVG unit in the tangential (azimuthal) direction. */
  tangentialScale(rho: number, radius: number): number {
    if (rho <= 1e-12) return this.radialScale(0.0, radius);
    let u = (rho * this.sinHalfCap) / this.rRef;
    if (u > 1.0) u = 1.0;
    const phi = 2.0 * Math.asin(u);
    return (radius * Math.sin(phi)) / rho;
  }

  /** Surface area (mm^2) per unit planar area at planar radius rho. */
  arealScale(rho: number, radius: number): number {
    return this.radialScale(rho, radius) * this.tangentialScale(rho, radius);
  }

  /** Return [min, mid, max] surface scale across the cap, for sizing. */
  scaleBounds(radius: number): [number, number, number] {
    let mn = Infinity;
    let mx = -Infinity;
    const N = 64;
    for (let i = 0; i < N; i++) {
      const rho = (this.rRef * i) / (N - 1);
      const r = this.radialScale(rho, radius);
      const t = this.tangentialScale(rho, radius);
      if (r < mn) mn = r;
      if (r > mx) mx = r;
      if (t < mn) mn = t;
      if (t > mx) mx = t;
    }
    const mid = this.radialScale(this.rRef * 0.5, radius);
    return [mn, mid, mx];
  }
}

File: ball_stencil/mapping.py
"""Plane -> sphere parameterization (Lambert azimuthal equal-area).

The SVG centre maps to the dome pole (+z).  Radial distance in the plane maps
to polar angle via the equal-area relation, scaled so the design reference
radius lands exactly at the cap angle:

    rho = |p - center|
    u   = (rho / R_ref) * sin(cap/2)        # in [0,1]
    phi = 2 * asin(u)                       # polar angle from pole
    theta = atan2(dy, dx)                   # azimuth (dy flipped if FLIP_V)

    P(r) = r * (sin phi cos theta, sin phi sin theta, cos phi)

This is an analytic *parameterization*, not a projection of pre-placed 3D
geometry: every output point originates from spherical coordinates.  Because
rho proportional to sin(phi/2), the mapping preserves area up to the constant
(R_ref / sin(cap/2))^2, so the relative size of each splash is preserved.
"""

from __future__ import annotations

from dataclasses import dataclass
from math import sin, pi

import numpy as np


@dataclass
class Mapper:
    center: np.ndarray          # (2,) float64
    r_ref: float                # SVG units mapped to cap angle
    cap_angle_rad: float
    flip_v: bool

    @property
    def _sin_half_cap(self) -> float:
        return sin(self.cap_angle_rad / 2.0)

    def directions(self, pts_xy: np.ndarray) -> np.ndarray:
        """Unit sphere directions for planar points; returns (N,3) float64."""
        p = np.asarray(pts_xy, dtype=np.float64).reshape(-1, 2)
        d = p - self.center
        dx = d[:, 0]
        dy = -d[:, 1] if self.flip_v else d[:, 1]
        rho = np.hypot(dx, dy)
        theta = np.arctan2(dy, dx)
        u = np.clip(rho / self.r_ref * self._sin_half_cap, 0.0, 1.0)
        phi = 2.0 * np.arcsin(u)
        sphi = np.sin(phi)
        dirs = np.empty((len(p), 3), dtype=np.float64)
        dirs[:, 0] = sphi * np.cos(theta)
        dirs[:, 1] = sphi * np.sin(theta)
        dirs[:, 2] = np.cos(phi)
        return dirs

    def to_sphere(self, pts_xy: np.ndarray, radius: float) -> np.ndarray:
        """Map planar points onto the sphere of the given radius. (N,3)"""
        return self.directions(pts_xy) * radius

    # -- scale helpers (mm of surface arc per SVG unit) ----------------------
    def radial_scale(self, rho: float, radius: float) -> float:
        """d(arc length)/d(rho) at planar radius ``rho``."""
        s = self._sin_half_cap / self.r_ref
        u = min(max(rho * s, 0.0), 1.0 - 1e-12)
        dphi_drho = 2.0 * s / np.sqrt(1.0 - u * u)
        return radius * dphi_drho

    def tangential_scale(self, rho: float, radius: float) -> float:
        """Arc length per SVG unit in the tangential (azimuthal) direction."""
        if rho <= 1e-12:
            return self.radial_scale(0.0, radius)
        u = min(rho * self._sin_half_cap / self.r_ref, 1.0)
        phi = 2.0 * np.arcsin(u)
        return radius * np.sin(phi) / rho

    def scale_bounds(self, radius: float) -> tuple[float, float, float]:
        """Return (min, mid, max) surface scale across the cap, for sizing."""
        samples = np.linspace(0.0, self.r_ref, 64)
        vals = []
        for rho in samples:
            vals.append(self.radial_scale(rho, radius))
            vals.append(self.tangential_scale(rho, radius))
        vals = np.asarray(vals)
        mid = self.radial_scale(self.r_ref * 0.5, radius)
        return float(vals.min()), float(mid), float(vals.max())


def build_mapper(center, r_ref: float, cap_angle_rad: float, flip_v: bool) -> Mapper:
    return Mapper(
        center=np.asarray(center, dtype=np.float64),
        r_ref=float(r_ref),
        cap_angle_rad=float(cap_angle_rad),
        flip_v=bool(flip_v),
    )
